DOCSP-25995 version reference guidelines (#142)
mongodb/docs-meta: source/style-guide/style/version.txt

.. _version-numbers:

> ==================

Reviewer: [nit]

Can we adjust the `=`'s here so they match the title length?

Author: Fixed

Version Numbers
==================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. |add| replace:: `versionadded <https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded>`__
.. |change| replace:: `versionchanged <https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionchanged>`__
.. |deprecated| replace:: `deprecated <https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated>`__

Versions are unique names assigned to product releases that are packaged 
with a set of features at a specific point in time. 

> MongoDB products follow semantic versioning (link to wikipedia). 

Reviewer 2: Issue:
Looks like the link is missing.

Suggestion:
Use the following software versioning wikipedia link:
https://en.wikipedia.org/w/index.php?title=Software_versioning&oldid=1147438049

  .. code-block:: rst

     .. versionadded:: 6.0

  If a version reference needs more explanation, add the content 
  to the directive block:

  .. code-block:: rst

     .. versionchanged:: 4.4

        This command returns the following output:

  If you need to call out multiple versions, set the most recent version
  first, then reference the others in a phrase afterwards.

  .. code-block:: rst

     .. versionadded:: 6.0 *(Also available in MongoDB 5.0.)*

  The italics marker matches the directive output.

  Snooty supports the following version directives: 

  .. hlist::
     :columns: 3

     * |add|
     * |change|
     * |deprecated|

* To refer to a version inline, include the product
  name before the version number, (for example, ``MongoDB 5.0``).  This ensures 
  that a search for version numbers correctly matches product and release.

* Do **not** reference versions freeform or with the phrase ``version X.X``.
  These references are much harder to identify programmatically as references 
  to a specific product version.

Identify EOL Releases
---------------------

You can find references to EOL releases of MonogDB using grep:

.. code-block:: shell

>    $ grep -RI \

Reviewer 2: Question: shouldn't you have the "n" option set if you want the line number output as shown in lines 110-112?

Author: Fixed

         -e ".. versionadded:: [0-9]" \
         -e ".. versionchanged:: [0-9]" \
         -e ".. deprecated:: [0-9]" \
         -e "MongoDB [0-9]" \
         source | \
         grep -v "6\.2" | \
         grep -v "6\.1" | \
         grep -v "6\.0" | \
         grep -v "5\.3" | \
         grep -v "5\.1" | \
         grep -v "5\.0" | \
         grep -v "4\.4" | \
         grep -v "4\.2" | \
         grep -v "source/release-notes"
   source/sharding.txt:128:- In MongoDB 4.0 and earlier, a document's shard key field value is
   source/sharding.txt:351:   Starting in MongoDB 4.0.3, setting up zones and zone ranges *before*
   source/sharding.txt:379:Starting in MongoDB 3.6, :doc:`change streams </changeStreams>` are
   ...

> This command could be added to a script and adjusted to work from an
> arbitrary directory:

Reviewer 3: ```suggestion
You can add the following command to a script and adjust it to work from an
arbitrary directory:
```

Author: Mitigated.

.. code-block:: bash

   grep -RIne ".. versionadded:: [0-9]" -e ".. versionchanged:: [0-9]" \
         -e ".. deprecated:: [0-9]" -e "MongoDB [0-9]" $1 | \
         grep -v "6\.2" | \
         grep -v "6\.1" | \
         grep -v "6\.0" | \
         grep -v "5\.3" | \
         grep -v "5\.1" | \
         grep -v "5\.0" | \
         grep -v "4\.4" | \
         grep -v "4\.2" | \
         grep -v "source/release-notes"

> To execute the script, call it with a path argument:

Reviewer 2: Suggestion:

I would omit this subsection on converting it to a script since it's mostly duplicate, and converting a shell command to a script can be learned via google.

There also seems to be some minor details missing on preparing the script. For example, I think you'll need to set execution permissions and to use a shebang interpreter directive to run the command as listed. It's also not apparent the script should be called "version.sh" nor that the "source" argument is the name of a relative directory.

Author: Fixed

.. _version-numbers:

==================
Version Numbers
==================

> .. default-domain:: mongodb

Author: Fixed

As you document features, you often need to reference a major or specific product release version to indicate when the feature became available, changed, or was deprecated.

> Version references can help explain product behavior

Reviewer: [suggestion, minimalism]

Remove "can help".

Author: Removed

related to a specific product release version.

With the exception of release notes and changelogs, which serve as historical
documents, avoid referencing product release versions that we no longer support.

> For examples, when a release of MongoDB Server reaches End of Life (EOL), remove references to that release.

Reviewer: [typo]

examples > example

Author: Fixed

> You can omit information about the following items:

Reviewer: [suggestion]

I think this sentence could be edited so it more closely ties to the previous paragraph. Consider:

"You can omit the following release-specific references: "

Author: Fixed

  .. code-block:: rst

     .. versionchanged:: 4.4

        Starting in MongoDB 4.4, this command returns the following output:

  Snooty supports three version directives: ``versionadded``, 
  ``versionchanged``, and ``deprecated``.

  .. note::
      
     Include **only** a single version number in this directive. Multiple
     version numbers render incorrectly in HTML.

* To refer to a version inline, include the product
  name before the version number, (for example, ``MongoDB 5.0``).  This ensures that
  a search for version numbers correctly matches product and release.

* Do **not** reference versions freeform or with the phrase ``version X.X``.
  These references are much harder to identify programmatically as references to 
  a specific product version.

Identify EOL Releases
---------------------

`versc <https://github.com/kennethdyer/versc>`__ is a tool for finding references
to EOL releases of products.  It operates similar to grep, performing a 
series of regular expression matches on each line of each file, searching for
{+rst+} version directives and ``PRODUCT X.Y`` pairings. It then
extracts the version references from lines it finds and evaluates whether these

> are legitimate version references versus references to unknown (read: EOL) releases.

Reviewer: [simplicity]

Can we simplify the last phrase to: "legitimate version references versus EOL release references", or something similar?

I find the `read: <x>` structure a little clunky.

Author: Section removed.

.. _version-reference:

==================
Version Reference

> ==================

Reviewer 2: Suggestion:
Maybe "Version Numbers" or "Versions" would match the rest of the page titles in the Style Guidelines section.

> ==================

Author: Fixed

.. default-domain:: mongodb

Documentation of features often requires reference to a specific or at least
major release of the product to indicate when the feature became available,
when the feature was changed, and when it was deprecated.

Readers find this information useful as it can explain unexpected behavior in
cases where they attempt to perform an action that is not possible on their
specific release of the product.

> With the exception of release notes and changelogs, which serve as a historical

Author: Fixed

> document, we should not reference releases of products that we no longer support.

Author: Fixed

That is, when a release of MongoDB Server reaches EOL,  references to this
release should be removed.

* Users don't need to know a feature was introduced in a release that is EOL,
  since it suggests a text that is out of date.

* Users don't need to know about features that were removed in a release 
  that is EOL, since it means the feature is unavailable in all supported 
  releases.

* Users don't need to know about features that were changed in a release
  that is EOL, since the change now applies to all available releases.

> This topic provides guidelines on how to reference product versions in order to
> make it easier to find these references once the release reaches EOL.

Reviewer 3: ```suggestion
Use the following guidelines to reference product versions so that
you can more easily find these references when the release reaches EOL.
```

General Guidelines
------------------

> When referencing a particular release of a product:

Reviewer 3: **Style:** See [Clarify Gerunds and Participles](https://www.mongodb.com/docs/meta/style-guide/writing/clarify-gerunds-participles/)

```suggestion
When you reference a specific product release version:
```

> * Prefer version directives wherever possible, as this is the easiest target

Author: Fixed

>   to later find and evaluate:

Author: Fixed

> * Prefer version directives wherever possible, as this is the easiest target
>   to later find and evaluate:
> 
>   .. code-block:: rst
> 
>      .. versionadded:: 6.0
> 

Reviewer 3: ```suggestion
* Use version directives wherever possible for findability:

  .. code-block:: rst

     .. versionadded:: 6.0

```

>   In cases where the reference requires further explanation, add the content 
>   to the directive block:
> 
>   .. code-block:: rst
> 
>      .. versionchanged:: 4.4
> 
>         Starting in MongoDB 4.4, this command returns the following output:

Reviewer 3: ```suggestion
  If a version reference needs more explanation, add the content 
  to the directive block:

  .. code-block:: rst

     .. versionchanged:: 4.4

        Starting in MongoDB 4.4, this command returns the following output:
```

>   Snooty supports three version directives: ``versionadded``, 

Author: Fixed

>   ``versionchanged``, and ``deprecated``

Reviewer 2: Suggestion:
Link these to material that explains what these are used for if that exists, or otherwise provide a short explanation of when to use them since this page is recommending using them.

Author: Fixed

  .. note::
      
     Do **not** include multiple version numbers in this directive, as it
     does not render correctly in HTML.

> * In cases where you need to refer to a version inline, include the product
>   name before the version number, (e.g. ``MongoDB 5.0``).  This ensures that
>   a search for version numbers will correctly match product and release for
>   checks.

Reviewer 3: ```suggestion
* To refer to a version inline, include the product
  name before the version number, (for example, ``MongoDB 5.0``).  This ensures that
  a search for version numbers correctly matches product and release.
```

* Do **not** reference versions freeform or with the phrase ``version X.X``
  as these are much harder to identify programmatically as references to 
  a specific product version.

Identifying EOL Releases
------------------------

> `versc <https://github.com/kennethdyer/versc>`__ is a tool for finding references

Author: Fixed

> to EOL releases of products.  It operates similar to grep, performing a 

Reviewer 2: Suggestion:
If I understand the functionality of the tool, I think it could be useful to identify whether these are inline text references or version directive references (or both).

Author: I'm in early stages of redesigning versc to match products to numbers and find other cases that might be hiding in the text.  Until that's ready, I've replaced this section with an example using grep.

series of Regular Expression matches on each line of each file looking for
reStructuredText version directives and ``PRODUCT X.Y`` pairings. It then
extracts the version references from lines it finds and evaluates whether these
are legitimate version references versus references to unknown (read: EOL) releases.

For example:

.. code-block:: console

   $ versc source/core/ | grep 2d.txt
   2.2   -   source/core/2d.txt:22   <geospatial-legacy>` from MongoDB 2.2 or earlier, *and*
   2.2   -   source/core/2d.txt:17   <geospatial-legacy>` used in MongoDB 2.2 and earlier.

This output is not exhaustive (versc doesn't parse the reStructuedText, so it 
isn't reading include files), but does provide a starting point in identifying

> references to EOL releases.

Reviewer 3: `This output is not exhaustive because versc doesn't parse the {+rst+}, so it isn't reading included files. However, versc does help you initially identify
references to EOL releases.`

Reviewer: I like @corryroot's suggestion here, I might just also consider shortening "versc does help you initially identify" to "versc helps initially identify..."

Author: Removed section in favor of grep example.
